src/core/vector/qgsvectorlayerrenderer.cpp

```diff
@@ -354,8 +355,26 @@ bool QgsVectorLayerRenderer::renderInternal( QgsFeatureRenderer *renderer, int r
     if ( mDiagramProvider )
       mDiagramProvider->setClipFeatureGeometry( mLabelClipFeatureGeom );
   }
+
+
   renderer->modifyRequestExtent( requestExtent, context );
 
+  double maximumExtentBuffer = renderer->maximumExtentBuffer( context );
+
+  if ( maximumExtentBuffer != 0 )
+  {
+    bool bufferDisappearsExtent = maximumExtentBuffer < 0 && ( requestExtent.width() + ( maximumExtentBuffer * 2 ) < 0 || requestExtent.height() + ( maximumExtentBuffer * 2 ) < 0 );
+
+    // nothing to draw
+    if ( bufferDisappearsExtent )
+    {
+      renderer->stopRender( context );
+      return true;
+    }
+
+    requestExtent = requestExtent.buffered( maximumExtentBuffer );
+  }
```

Comment:
It might be cleaner to just move this logic into  QgsFeatureRenderer::modifyRequestExtent ?